static/js/admin/userprofile.js

```diff
@@ -0,0 +1,113 @@
+document.addEventListener('DOMContentLoaded', () => {
+  'use strict';
+
+  function are_set_equal(a, b) {
+    if (a.size !== b.size) {
+      return false;
+    }
+    for (const value of a) {
+      if (!b.has(value)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  function get_search_bar_terms() {
+    /**
+     * Return a Set of all search terms in the search bar.
+     */
+    const search_term = search_bar.value.trim();
+    return new Set(search_term ? search_term.split(',') : []);
+  }
+
+  function highlight_ips_not_in(values) {
+    /**
+     * Highlight fields containing an IP not already in Set passed as argument
+     * by setting a 'notinsearch' class. Also reset it on existing fields if
+     * it was present but is no longer accurate, and apply the 'dirty' class
+     * to the search bar if the value is different from when the page was
+     * loaded.
+     */
+    for (const field of ip_fields) {
+      const value = field.textContent;
+      if (value.length <= 1) {
+        continue;
+      }
+      if (!values.has(value)) {
+        field.classList.add('notinsearch');
+      } else {
+        field.classList.remove('notinsearch');
+      }
+    }
+
+    if (!are_set_equal(original_search_terms, values)) {
+      search_bar.classList.add('dirty');
+    } else {
+      search_bar.classList.remove('dirty');
+    }
+  }
+
+  function add_add_remove_links() {
+    /**
+     * Add hidden links next to every IP field to add or remove the value to
+     * the search terms.
+     * They will be shown when hovering elements with the 'hasadremovedip'
+     * class, so add that class to each parent as well.
+     */
+    for (const field of ip_fields) {
+      const value = field.textContent;
+      if (value.length <= 1) {
+        continue;
+      }
+      const add_link = document.createElement('a');
+      add_link.className = 'addlink';
+      field.appendChild(add_link);
+
+      const delete_link = document.createElement('a');
+      delete_link.className = 'deletelink';
+      field.appendChild(delete_link);
+
+      field.classList.add('hasaddremoveip');
+    }
+  }
+
+  const result_list = document.querySelector(
+    'body.change-list.model-userprofile.change-list #result_list',
+  );
+
+  if (!result_list) {
+    // This is only for the userprofile result change list page.
+    return;
+  }
+
+  const search_bar = document.querySelector('#searchbar');
+  const original_search_terms = get_search_bar_terms();
+
+  const ip_fields = document.querySelectorAll(
+    '.field-last_login_ip, \
+    .field-restriction_history__last_login_ip, \
+    .field-restriction_history__ip_address, \
+    .field-_ratings_all__ip_address, \
+    .field-activitylog__iplog__ip_address',
+  );
```

Comment:
can we move `search_bar` and `ip_fields` `const`ants to the top of the file if they're going to referenced inside the functions - it makes the code hard to follow

Reply:
I wanted to initialize them only if needed, so after the `result_list` check. And being `const` I need to initialize them right away... The code is short & simple enough that I can make them `var`s and move the declaration up though.